src/npctalk.cpp

```diff
@@ -2792,6 +3193,151 @@ void talk_function::give_equipment(npc *p)
     p->add_effect("asked_for_item", 1800);
 }
 
+void talk_function::give_aid(npc *p)
+{
+    g->u.cash -= 20000;
+    p->add_effect("currently_busy", 300);
+    body_part bp_healed;
+    for (int i = 0; i < num_hp_parts; i++) {
+        bp_healed = player::hp_to_bp( hp_part(i) );
+        g->u.heal(hp_part(i), 5*rng(2,5));
+        if (g->u.has_effect("bite", bp_healed)) {
+            g->u.remove_effect("bite", bp_healed);
+        }
+        if (g->u.has_effect("bleed", bp_healed)) {
+            g->u.remove_effect("bleed", bp_healed);
+        }
+        if (g->u.has_effect("infected", bp_healed)) {
+            g->u.remove_effect("infected", bp_healed);
+        }
+    }
+    g->u.assign_activity(ACT_WAIT, 10000);
+}
+
+void talk_function::give_all_aid(npc *p)
+{
+    g->u.cash -= 30000;
+    p->add_effect("currently_busy", 300);
+    give_aid(p);
+    body_part bp_healed;
+    for( auto &elem : g->active_npc ) {
+        if (rl_dist( elem->pos(), g->u.pos() ) < PICKUP_RANGE && elem->is_friend()){
+            for (int i = 0; i < num_hp_parts; i++) {
+                bp_healed = player::hp_to_bp( hp_part(i) );
+                elem->heal(hp_part(i), 5*rng(2,5));
+                if (elem->has_effect("bite", bp_healed)) {
+                    elem->remove_effect("bite", bp_healed);
+                }
+                if (elem->has_effect("bleed", bp_healed)) {
+                    elem->remove_effect("bleed", bp_healed);
+                }
+                if (elem->has_effect("infected", bp_healed)) {
+                    elem->remove_effect("infected", bp_healed);
+                }
+            }
+        }
+    }
+}
+
+void talk_function::construction_tips(npc *p)
+{
+    g->u.cash -= 2000;
+    g->u.practice("carpentry", 30);
+    g->u.assign_activity(ACT_WAIT, 600);
+    p->add_effect("currently_busy", 600);
+}
+
+void talk_function::buy_beer(npc *p)
+{
+    item cont = item("bottle_glass", 0, false);
+    item liq = item("beer", 0);
```

Comment:
I'm not sure what you need more (approximation of what?). If you need to set the charges of the liquid, just do it after it has been put into the container (or put it there manually):
```C++
item cont( "bottle", 0 );
cont.put_in( liq( "beer", 0 ) );
cont.content[0].charges = 10;
g->u.i_add( cont );

// Or set the charges before:
item liq( "beer", 0 );
item cont( "bottle", 0 );
liq.charges = 10;
cont.put_in( beer );
g->u.i_add( cont );

// You can use the contents vector direcly instead of put_in:
item cont( "bottle", 0 );
cont.contents.push_back( liq( "beer", 0 ) );
cont.content[0].charges = 10;
g->u.i_add( cont );
```

Anything to avoid `handle_liquid`. That function is used to allow the player to choose the target container. Putting things into a container is *not* its purpose. You can put things direcly into the item using the item class.

----

If you want to avoid giving the player a free bottle, just call `move_liquid` and check the remaining charges (used that way in `player::siphon`, for example):
```C++
item liq( "beer", 0 );
const long default_charges = 10
liq.charges = default_charges;
const int remain = g->move_liquid( liq );
if( remain <= -1 ) {
    // canceled, nothing has been used.
    return;
}
const long used_charges = default_charges - remain;
// Note: used_charges may be 0, but that's fine here.
g->u.cash -= used_charges * price;
```

You can check for enough container space with `player::has_container_for` (the crafting code does this).